File: reference-implementation/derive_key.py
#!/usr/bin/env python3
"""A small, standalone reference implementation of the XILENT-KEY-V1 derivation.

This program intentionally keeps the full derivation in one readable file. It
uses only Python's standard library, so it can also serve as a recovery aid.

Usage:
    python derive_key.py path/to/master-key.mkf object:backup:v1
    python derive_key.py --verify-test-vector
    python derive_key.py --write-test-vector-markdown vectors.json vectors.md
    python derive_key.py --verify-test-vectors vectors.json
"""

# Standard-library modules provide the cryptographic primitives and command-line
# interface. No third-party package is required.
import argparse
import hashlib
import json
import hmac
import re
import secrets
import unicodedata
from getpass import getpass
from pathlib import Path


# This section defines the complete, versioned algorithm contract. Do not change
# any value here if a key must remain compatible with XILENT-KEY-V1.
PBKDF2_ITERATION_COUNT = 600_000
MASTER_KEY_FILE_LENGTH_BYTES = 48
MASTER_KEY_MATERIAL_LENGTH_BYTES = 32
MASTER_KEY_SALT_LENGTH_BYTES = 16
DERIVED_KEY_LENGTH_BYTES = 32
PASSWORD_SALT_PREFIX = b"XILENT-PBKDF2-V1\0"
HKDF_SALT_PREFIX = b"XILENT-HKDF-V1\0"
HKDF_INFO_PREFIX = b"XILENT-KEY-V1\0"
INFO_STRING_PATTERN = re.compile(r"^[a-z0-9][a-z0-9:._-]{0,127}$", re.ASCII)


def derive_key_from_master_key_file(
    master_key_file_path: Path,
    memorized_secret: str,
    info_string: str,
) -> str:
    """Read one .mkf file and return its 64-character lowercase hex key."""

    # An .mkf contains exactly 32 bytes of master-key material followed by 16
    # bytes of salt. Reading it here makes the command-line flow explicit.
    master_key_file_contents = master_key_file_path.read_bytes()
    return derive_key_from_master_key_bytes(
        master_key_file_contents,
        memorized_secret,
        info_string,
    )


def derive_key_from_master_key_bytes(
    master_key_file_contents: bytes,
    memorized_secret: str,
    info_string: str,
) -> str:
    """Derive a key from the 48-byte .mkf contents and the two user inputs."""

    # Reject malformed key files before splitting them, so their layout never
    # becomes ambiguous.
    if len(master_key_file_contents) != MASTER_KEY_FILE_LENGTH_BYTES:
        raise ValueError(
            "The master-key file must contain exactly 48 bytes: "
            "32 bytes of key material followed by 16 bytes of salt."
        )

    # A blank secret provides no second factor and is invalid. Whitespace is
    # significant, so it is deliberately not stripped.
    if not memorized_secret:
        raise ValueError("The memorized secret must not be empty.")

    # The info string is an identifier, not free-form text. Its restricted ASCII
    # form avoids encoding ambiguity and makes recovery records predictable.
    if not INFO_STRING_PATTERN.fullmatch(info_string):
        raise ValueError(
            "The info string must be 1-128 lowercase ASCII characters, start "
            "with a-z or 0-9, and otherwise use only a-z, 0-9, :, ., _, or -."
        )

    # Split the fixed file layout. These slices are the two binary inputs stored
    # in the .mkf file and must be used in this order.
    master_key_material = master_key_file_contents[:MASTER_KEY_MATERIAL_LENGTH_BYTES]
    master_key_salt = master_key_file_contents[MASTER_KEY_MATERIAL_LENGTH_BYTES:]

    # Normalize equivalent Unicode spellings (for example, café entered as one
    # code point or as e plus an accent) before UTF-8 encodes the secret.
    normalized_memorized_secret_bytes = unicodedata.normalize(
        "NFC", memorized_secret
    ).encode("utf-8")

    # PBKDF2 turns the memorized secret into a 32-byte key. Prefixing the file
    # salt labels this use of the salt and separates it from the HKDF use below.
    password_derivation_salt = PASSWORD_SALT_PREFIX + master_key_salt
    memorized_secret_key = hashlib.pbkdf2_hmac(
        "sha256",
        normalized_memorized_secret_bytes,
        password_derivation_salt,
        PBKDF2_ITERATION_COUNT,
        dklen=DERIVED_KEY_LENGTH_BYTES,
    )

    # HKDF receives both independent secret inputs in a fixed order: the stored
    # 32-byte master-key material first, then the PBKDF2 result.
    hkdf_input_key_material = master_key_material + memorized_secret_key
    hkdf_salt = HKDF_SALT_PREFIX + master_key_salt
    hkdf_info = HKDF_INFO_PREFIX + info_string.encode("ascii")
    derived_key_bytes = hkdf_sha256(
        hkdf_input_key_material,
        hkdf_salt,
        hkdf_info,
        DERIVED_KEY_LENGTH_BYTES,
    )

    # Lowercase hexadecimal is the portable output representation and is what
    # the application returns. Thirty-two bytes therefore become 64 characters.
    return derived_key_bytes.hex()


def hkdf_sha256(
    input_key_material: bytes,
    salt: bytes,
    info: bytes,
    output_length_bytes: int,
) -> bytes:
    """Implement RFC 5869 HKDF with SHA-256 using explicit Extract and Expand."""

    # SHA-256 has a 32-byte digest, and HKDF permits at most 255 digest blocks.
    sha256_digest_length_bytes = hashlib.sha256().digest_size
    if not 0 <= output_length_bytes <= 255 * sha256_digest_length_bytes:
        raise ValueError("The requested HKDF output length is invalid.")

    # Extract: use the salt as the HMAC key to convert input material into a
    # fixed-length pseudorandom key (PRK).
    pseudorandom_key = hmac.digest(salt, input_key_material, "sha256")

    # Expand: each block incorporates the preceding block, the context-specific
    # info bytes, and a one-byte counter. This derivation needs only one block.
    output = bytearray()
    previous_block = b""
    for block_counter in range(1, 256):
        previous_block = hmac.digest(
            pseudorandom_key,
            previous_block + info + bytes([block_counter]),
            "sha256",
        )
        output.extend(previous_block)
        if len(output) >= output_length_bytes:
            return bytes(output[:output_length_bytes])

    # The range check above makes this unreachable; it documents the invariant.
    raise AssertionError("HKDF expansion exceeded its maximum block count.")


def verify_public_test_vector() -> None:
    """Verify the original single interoperability vector."""

    public_test_master_key_file_contents = bytes(range(32)) + bytes(range(16))
    expected_derived_key = (
        "75f76fdedf2d5384a93c94440c9f6a731ecb72ea896970811be909a7a61a3a37"
    )
    actual_derived_key = derive_key_from_master_key_bytes(
        public_test_master_key_file_contents,
        "public test memory secret",
        "cryptotest:test:v1",
    )
    if not secrets.compare_digest(actual_derived_key, expected_derived_key):
        raise AssertionError("The public test vector did not match.")
    print(f"PASS cryptotest:test:v1: {actual_derived_key}")


def write_test_vectors_markdown(vector_path: Path, markdown_path: Path) -> None:
    """Export the shared JSON vectors as readable, non-secret documentation."""

    vectors = json.loads(vector_path.read_text(encoding="utf-8"))
    lines = [
        "# XILENT-KEY-V1 Reference Test Vectors",
        "",
        "Generated from `tests/Xilent.KeyDeriver.Tests/TestVectors/derivation-vectors.json`.",
        "These are public interoperability vectors; the memorized secrets are test data only.",
        "",
        "| # | Master-key bytes (hex) | Memorized secret | Info | Expected derived key |",
        "|---:|---|---|---|---|",
    ]
    for index, vector in enumerate(vectors, start=1):
        lines.append(
            f"| {index} | `{vector['mkfHex']}` | "
            f"`{vector['memorySecret']}` | `{vector['info']}` | "
            f"`{vector['expected']}` |"
        )
    lines.extend([
        "",
        "## Algorithm parameters",
        "",
        f"- PBKDF2 iterations: `{PBKDF2_ITERATION_COUNT}`",
        f"- Master-key material length: `{MASTER_KEY_MATERIAL_LENGTH_BYTES}` bytes",
        f"- Master-key salt length: `{MASTER_KEY_SALT_LENGTH_BYTES}` bytes",
        f"- Derived key length: `{DERIVED_KEY_LENGTH_BYTES}` bytes",
        f"- Password/PBKDF2 salt prefix: `{PASSWORD_SALT_PREFIX!r}`",
        f"- HKDF salt prefix: `{HKDF_SALT_PREFIX!r}`",
        f"- HKDF info prefix: `{HKDF_INFO_PREFIX!r}`",
        "",
    ])
    markdown_path.write_text("\n".join(lines), encoding="utf-8")
    print(f"Test vectors copied to {markdown_path}")


def verify_test_vectors(vector_path: Path) -> bool:
    """Verify every vector shared with the C# test project."""

    vectors = json.loads(vector_path.read_text(encoding="utf-8"))
    print("=== XILENT-KEY-V1 Reference Verification ===")
    print("Expected derived key: listed per vector below.")
    print(f"Prefixes: pws={PASSWORD_SALT_PREFIX!r}, hkdf={HKDF_SALT_PREFIX!r}, info={HKDF_INFO_PREFIX!r}")
    print(f"Parameters: iterations={PBKDF2_ITERATION_COUNT}, key length={MASTER_KEY_MATERIAL_LENGTH_BYTES} bytes, salt length={MASTER_KEY_SALT_LENGTH_BYTES} bytes")
    print(f"Context: {len(vectors)} shared C# test vectors from {vector_path}")
    print()
    all_passed = True
    for index, vector in enumerate(vectors, start=1):
        actual = derive_key_from_master_key_bytes(
            bytes.fromhex(vector["mkfHex"]), vector["memorySecret"], vector["info"]
        )
        passed = secrets.compare_digest(actual, vector["expected"])
        all_passed = all_passed and passed
        result = "PASS" if passed else "FAIL"
        print(f"{result} vector {index} ({vector['info']})")
        print(f"  expected derived key: {vector['expected']}")
        print(f"  generated derived key: {actual}")
    print(f"\nOverall result: {'PASS' if all_passed else 'FAIL'}")
    return all_passed


def main() -> None:
    """Parse the intentionally small command-line interface and derive a key."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("master_key_file", type=Path, nargs="?")
    parser.add_argument("info_string", nargs="?")
    parser.add_argument("--verify-test-vector", action="store_true")
    parser.add_argument("--verify-test-vectors", type=Path)
    parser.add_argument("--write-test-vector-markdown", nargs=2, metavar=("JSON", "MARKDOWN"), type=Path)
    arguments = parser.parse_args()

    if arguments.write_test_vector_markdown:
        write_test_vectors_markdown(*arguments.write_test_vector_markdown)
        return

    # The test-vector modes need no key file and never process user secrets.
    if arguments.verify_test_vectors is not None:
        raise SystemExit(0 if verify_test_vectors(arguments.verify_test_vectors) else 1)
    if arguments.verify_test_vector:
        verify_public_test_vector()
        return

    # Require both positional values together so a partial command cannot run.
    if arguments.master_key_file is None or arguments.info_string is None:
        parser.error("master_key_file and info_string are required unless verifying.")

    # getpass prevents the secret from being echoed. The secret is never written
    # to disk by this program, although Python cannot guarantee memory erasure.
    memorized_secret = getpass("Memorized secret: ")
    print(
        derive_key_from_master_key_file(
            arguments.master_key_file,
            memorized_secret,
            arguments.info_string,
        )
    )


if __name__ == "__main__":
    main()

File: reference-implementation/derive_key.ts
#!/usr/bin/env tsx
/**
 * A small, standalone TypeScript reference implementation of XILENT-KEY-V1.
 *
 * This intentionally keeps the complete derivation in one readable file. It
 * uses Node.js built-in modules only, so it can serve as recovery documentation.
 *
 * Usage:
 *   npx tsx derive_key.ts path/to/master-key.mkf object:backup:v1
 *   npx tsx derive_key.ts --verify-test-vector
 *   npx tsx derive_key.ts --write-test-vector-markdown vectors.json vectors.md
 *   npx tsx derive_key.ts --verify-test-vectors vectors.json
 */

// Node.js built-in modules provide the cryptographic primitives, file access,
// and command-line input. No third-party package is required by this script.
import {
    createHmac,
    pbkdf2Sync,
    timingSafeEqual,
} from "node:crypto";
import { readFileSync, writeFileSync } from "node:fs";
import { basename } from "node:path";
import { createInterface } from "node:readline/promises";
import { stdin, stdout } from "node:process";

// This section defines the complete, versioned algorithm contract. Do not
// change any value here if a key must remain compatible with XILENT-KEY-V1.
const PBKDF2_ITERATION_COUNT: number = 600_000;
const MASTER_KEY_FILE_LENGTH_BYTES: number = 48;
const MASTER_KEY_MATERIAL_LENGTH_BYTES: number = 32;
const MASTER_KEY_SALT_LENGTH_BYTES: number = 16;
const DERIVED_KEY_LENGTH_BYTES: number = 32;
const PASSWORD_SALT_PREFIX: Buffer = Buffer.from("XILENT-PBKDF2-V1\0", "utf8");
const HKDF_SALT_PREFIX: Buffer = Buffer.from("XILENT-HKDF-V1\0", "utf8");
const HKDF_INFO_PREFIX: Buffer = Buffer.from("XILENT-KEY-V1\0", "utf8");
const INFO_STRING_PATTERN: RegExp = /^[a-z0-9][a-z0-9:._-]{0,127}$/;

/** One shared XILENT-KEY-V1 interoperability vector. */
class TestVector {
    public readonly masterKeyFileHex: string;
    public readonly memorizedSecret: string;
    public readonly infoString: string;
    public readonly expectedDerivedKey: string;

    public constructor(
        masterKeyFileHex: string,
        memorizedSecret: string,
        infoString: string,
        expectedDerivedKey: string,
    ) {
        this.masterKeyFileHex = masterKeyFileHex;
        this.memorizedSecret = memorizedSecret;
        this.infoString = infoString;
        this.expectedDerivedKey = expectedDerivedKey;
    }
}

/** The JSON shape shared with the C# test project. */
type RawTestVector = {
    mkfHex: string;
    memorySecret: string;
    info: string;
    expected: string;
};

/** Read the shared JSON vectors into named values. */
function loadTestVectors(vectorPath: string): TestVector[] {
    const fileText: string = readFileSync(vectorPath, "utf8");
    const rawVectors: RawTestVector[] = JSON.parse(fileText) as RawTestVector[];
    const vectors: TestVector[] = [];

    for (const rawVector of rawVectors) {
        const vector: TestVector = new TestVector(
            rawVector.mkfHex,
            rawVector.memorySecret,
            rawVector.info,
            rawVector.expected,
        );
        vectors.push(vector);
    }

    return vectors;
}

/** Read one .mkf file and return its 64-character lowercase hexadecimal key. */
function deriveKeyFromMasterKeyFile(
    masterKeyFilePath: string,
    memorizedSecret: string,
    infoString: string,
): string {
    // An .mkf contains 32 bytes of master-key material followed by 16 bytes of
    // salt. Reading it here makes the command-line flow explicit.
    const masterKeyFileContents: Buffer = readFileSync(masterKeyFilePath);
    return deriveKeyFromMasterKeyBytes(
        masterKeyFileContents,
        memorizedSecret,
        infoString,
    );
}

/** Derive a key from the 48-byte .mkf contents and two user inputs. */
function deriveKeyFromMasterKeyBytes(
    masterKeyFileContents: Buffer,
    memorizedSecret: string,
    infoString: string,
): string {
    // Reject malformed key files before splitting them, so their layout never
    // becomes ambiguous.
    if (masterKeyFileContents.length !== MASTER_KEY_FILE_LENGTH_BYTES) {
        throw new Error(
            "The master-key file must contain exactly 48 bytes: " +
            "32 bytes of key material followed by 16 bytes of salt.",
        );
    }

    // A blank secret provides no second factor and is invalid. Whitespace is
    // significant, so it is deliberately not trimmed.
    if (memorizedSecret.length === 0) {
        throw new Error("The memorized secret must not be empty.");
    }

    // The info string is an identifier, not free-form text. Its restricted ASCII
    // form avoids encoding ambiguity and makes recovery records predictable.
    if (!INFO_STRING_PATTERN.test(infoString)) {
        throw new Error(
            "The info string must be 1-128 lowercase ASCII characters, start " +
            "with a-z or 0-9, and otherwise use only a-z, 0-9, :, ., _, or -.",
        );
    }

    // Split the fixed file layout. These slices are the two binary inputs stored
    // in the .mkf file and must be used in this order.
    const masterKeyMaterial: Buffer = masterKeyFileContents.subarray(
        0,
        MASTER_KEY_MATERIAL_LENGTH_BYTES,
    );
    const masterKeySalt: Buffer = masterKeyFileContents.subarray(
        MASTER_KEY_MATERIAL_LENGTH_BYTES,
        MASTER_KEY_FILE_LENGTH_BYTES,
    );

    // Normalize equivalent Unicode spellings before UTF-8 encodes the secret.
    const normalizedMemorizedSecret: string = memorizedSecret.normalize("NFC");
    const normalizedMemorizedSecretBytes: Buffer = Buffer.from(
        normalizedMemorizedSecret,
        "utf8",
    );

    // PBKDF2 turns the memorized secret into a 32-byte key. Prefixing the file
    // salt labels this use of the salt and separates it from the HKDF use below.
    const passwordDerivationSalt: Buffer = Buffer.concat([
        PASSWORD_SALT_PREFIX,
        masterKeySalt,
    ]);
    const memorizedSecretKey: Buffer = pbkdf2Sync(
        normalizedMemorizedSecretBytes,
        passwordDerivationSalt,
        PBKDF2_ITERATION_COUNT,
        DERIVED_KEY_LENGTH_BYTES,
        "sha256",
    );

    // HKDF receives both independent secret inputs in a fixed order: stored
    // master-key material first, then the PBKDF2 result.
    const hkdfInputKeyMaterial: Buffer = Buffer.concat([
        masterKeyMaterial,
        memorizedSecretKey,
    ]);
    const hkdfSalt: Buffer = Buffer.concat([HKDF_SALT_PREFIX, masterKeySalt]);
    const infoStringBytes: Buffer = Buffer.from(infoString, "ascii");
    const hkdfInfo: Buffer = Buffer.concat([HKDF_INFO_PREFIX, infoStringBytes]);
    const derivedKeyBytes: Buffer = hkdfSha256(
        hkdfInputKeyMaterial,
        hkdfSalt,
        hkdfInfo,
        DERIVED_KEY_LENGTH_BYTES,
    );

    // Lowercase hexadecimal is the portable output representation. Thirty-two
    // bytes therefore become 64 characters.
    return derivedKeyBytes.toString("hex");
}

/** Implement RFC 5869 HKDF with SHA-256 using explicit Extract and Expand. */
function hkdfSha256(
    inputKeyMaterial: Buffer,
    salt: Buffer,
    info: Buffer,
    outputLengthBytes: number,
): Buffer {
    // SHA-256 has a 32-byte digest, and HKDF permits at most 255 digest blocks.
    const sha256DigestLengthBytes: number = 32;
    const maximumOutputLengthBytes: number = 255 * sha256DigestLengthBytes;
    if (outputLengthBytes < 0 || outputLengthBytes > maximumOutputLengthBytes) {
        throw new Error("The requested HKDF output length is invalid.");
    }

    // Extract: use the salt as the HMAC key to produce a pseudorandom key.
    const pseudorandomKey: Buffer = hmacSha256(salt, inputKeyMaterial);

    // Expand: each block incorporates the preceding block, context-specific
    // info bytes, and a one-byte counter. This derivation needs one block.
    let output: Buffer = Buffer.alloc(0);
    let previousBlock: Buffer = Buffer.alloc(0);
    for (let blockCounter: number = 1; blockCounter <= 255; blockCounter += 1) {
        const counterByte: Buffer = Buffer.from([blockCounter]);
        const blockInput: Buffer = Buffer.concat([
            previousBlock,
            info,
            counterByte,
        ]);
        previousBlock = hmacSha256(pseudorandomKey, blockInput);
        output = Buffer.concat([output, previousBlock]);

        if (output.length >= outputLengthBytes) {
            return output.subarray(0, outputLengthBytes);
        }
    }

    // The range check above makes this unreachable; it documents the invariant.
    throw new Error("HKDF expansion exceeded its maximum block count.");
}

/** Calculate one HMAC-SHA-256 value with explicit key and message names. */
function hmacSha256(key: Buffer, message: Buffer): Buffer {
    const hmac = createHmac("sha256", key);
    hmac.update(message);
    return hmac.digest();
}

/** Compare equal-length public strings using Node's timing-safe comparison. */
function stringsMatch(actual: string, expected: string): boolean {
    const actualBytes: Buffer = Buffer.from(actual, "utf8");
    const expectedBytes: Buffer = Buffer.from(expected, "utf8");
    return actualBytes.length === expectedBytes.length &&
        timingSafeEqual(actualBytes, expectedBytes);
}

/** Create public test bytes containing 00, 01, 02, through the requested length. */
function createSequentialTestBytes(length: number): Buffer {
    const bytes: Buffer = Buffer.alloc(length);
    for (let index: number = 0; index < length; index += 1) {
        bytes[index] = index;
    }
    return bytes;
}

/** Verify the original single interoperability vector. */
function verifyPublicTestVector(): void {
    const masterKeyMaterial: Buffer = createSequentialTestBytes(32);
    const masterKeySalt: Buffer = createSequentialTestBytes(16);
    const masterKeyFileContents: Buffer = Buffer.concat([
        masterKeyMaterial,
        masterKeySalt,
    ]);
    const expectedDerivedKey: string =
        "75f76fdedf2d5384a93c94440c9f6a731ecb72ea896970811be909a7a61a3a37";
    const actualDerivedKey: string = deriveKeyFromMasterKeyBytes(
        masterKeyFileContents,
        "public test memory secret",
        "cryptotest:test:v1",
    );

    if (!stringsMatch(actualDerivedKey, expectedDerivedKey)) {
        throw new Error("The public test vector did not match.");
    }

    console.log(`PASS cryptotest:test:v1: ${actualDerivedKey}`);
}

/** Export shared JSON vectors as readable, non-secret Markdown documentation. */
function writeTestVectorsMarkdown(vectorPath: string, markdownPath: string): void {
    const vectors: TestVector[] = loadTestVectors(vectorPath);
    const lines: string[] = [
        "# XILENT-KEY-V1 Reference Test Vectors",
        "",
        "Generated from `tests/Xilent.KeyDeriver.Tests/TestVectors/derivation-vectors.json`.",
        "These are public interoperability vectors; memorized secrets are test data only.",
        "",
        "| # | Master-key bytes (hex) | Memorized secret | Info | Expected derived key |",
        "|---:|---|---|---|---|",
    ];

    for (let index: number = 0; index < vectors.length; index += 1) {
        const vector: TestVector = vectors[index];
        const vectorNumber: number = index + 1;
        lines.push(
            `| ${vectorNumber} | \`${vector.masterKeyFileHex}\` | ` +
            `\`${vector.memorizedSecret}\` | \`${vector.infoString}\` | ` +
            `\`${vector.expectedDerivedKey}\` |`,
        );
    }

    lines.push(
        "",
        "## Algorithm parameters",
        "",
        `- PBKDF2 iterations: \`${PBKDF2_ITERATION_COUNT}\``,
        `- Master-key material length: \`${MASTER_KEY_MATERIAL_LENGTH_BYTES}\` bytes`,
        `- Master-key salt length: \`${MASTER_KEY_SALT_LENGTH_BYTES}\` bytes`,
        `- Derived key length: \`${DERIVED_KEY_LENGTH_BYTES}\` bytes`,
        `- Password/PBKDF2 salt prefix: \`XILENT-PBKDF2-V1\\0\``,
        `- HKDF salt prefix: \`XILENT-HKDF-V1\\0\``,
        `- HKDF info prefix: \`XILENT-KEY-V1\\0\``,
        "",
    );

    writeFileSync(markdownPath, lines.join("\n"), "utf8");
    console.log(`Test vectors copied to ${markdownPath}`);
}

/** Verify every vector shared with the C# test project. */
function verifyTestVectors(vectorPath: string): boolean {
    const vectors: TestVector[] = loadTestVectors(vectorPath);
    console.log("=== XILENT-KEY-V1 Reference Verification ===");
    console.log("Expected derived key: listed per vector below.");
    console.log("Prefixes: pws=XILENT-PBKDF2-V1\\0, hkdf=XILENT-HKDF-V1\\0, info=XILENT-KEY-V1\\0");
    console.log(`Parameters: iterations=${PBKDF2_ITERATION_COUNT}, key length=${MASTER_KEY_MATERIAL_LENGTH_BYTES} bytes, salt length=${MASTER_KEY_SALT_LENGTH_BYTES} bytes`);
    console.log(`Context: ${vectors.length} shared C# test vectors from ${vectorPath}`);
    console.log();

    let allPassed: boolean = true;
    for (let index: number = 0; index < vectors.length; index += 1) {
        const vector: TestVector = vectors[index];
        const masterKeyFileContents: Buffer = Buffer.from(vector.masterKeyFileHex, "hex");
        const actualDerivedKey: string = deriveKeyFromMasterKeyBytes(
            masterKeyFileContents,
            vector.memorizedSecret,
            vector.infoString,
        );
        const passed: boolean = stringsMatch(actualDerivedKey, vector.expectedDerivedKey);
        if (!passed) {
            allPassed = false;
        }

        const result: string = passed ? "PASS" : "FAIL";
        console.log(`${result} vector ${index + 1} (${vector.infoString})`);
        console.log(`  expected derived key: ${vector.expectedDerivedKey}`);
        console.log(`  generated derived key: ${actualDerivedKey}`);
    }

    console.log(`\nOverall result: ${allPassed ? "PASS" : "FAIL"}`);
    return allPassed;
}

/** Display the intentionally small command-line interface. */
function printUsage(): void {
    const executableName: string = basename(process.argv[1] ?? "derive_key.ts");
    console.error("Usage:");
    console.error(`  npx tsx ${executableName} <master-key-file> <info-string>`);
    console.error(`  npx tsx ${executableName} --verify-test-vector`);
    console.error(`  npx tsx ${executableName} --verify-test-vectors <vectors.json>`);
    console.error(`  npx tsx ${executableName} --write-test-vector-markdown <vectors.json> <vectors.md>`);
}

/** Prompt for a secret without writing it to disk. */
async function readMemorizedSecret(): Promise<string> {
    // Node's standard readline prompt echoes input. This script is reference
    // documentation rather than a hardened secret-entry tool; use XVault for
    // normal use where its platform protections are appropriate.
    const terminal = createInterface({ input: stdin, output: stdout });
    const memorizedSecret: string = await terminal.question("Memorized secret: ");
    terminal.close();
    return memorizedSecret;
}

/** Parse the small CLI and either verify vectors or derive one key. */
async function main(): Promise<void> {
    const arguments_: string[] = process.argv.slice(2);

    if (arguments_.length === 1 && arguments_[0] === "--verify-test-vector") {
        verifyPublicTestVector();
        return;
    }

    if (arguments_.length === 2 && arguments_[0] === "--verify-test-vectors") {
        const vectorsPassed: boolean = verifyTestVectors(arguments_[1]);
        process.exitCode = vectorsPassed ? 0 : 1;
        return;
    }

    if (
        arguments_.length === 3 &&
        arguments_[0] === "--write-test-vector-markdown"
    ) {
        writeTestVectorsMarkdown(arguments_[1], arguments_[2]);
        return;
    }

    // Require both positional values together so a partial command cannot run.
    if (arguments_.length !== 2) {
        printUsage();
        process.exitCode = 2;
        return;
    }

    const masterKeyFilePath: string = arguments_[0];
    const infoString: string = arguments_[1];
    const memorizedSecret: string = await readMemorizedSecret();
    const derivedKey: string = deriveKeyFromMasterKeyFile(
        masterKeyFilePath,
        memorizedSecret,
        infoString,
    );
    console.log(derivedKey);
}

// Report errors consistently while retaining a non-zero process exit status.
main().catch((error: unknown) => {
    const message: string = error instanceof Error ? error.message : String(error);
    console.error(`Error: ${message}`);
    process.exitCode = 1;
});
